### PR TITLE
Fix sw_psycopg2 register_type()

### PR DESCRIPTION
**Pull request by tom-pytel**
`psycopg2._psycopg.register_type()` can be called with a connection or cursor, but since these are wrapped proxy objects when instrumented the binary module `_psycopg` does not recognize them causing calls like `psycopg2.extras.register_json()` to fail:
```
...
   File "/usr/local/lib/python3.10/site-packages/psycopg2/_json.py", line 120, in register_json
     register_type(JSON, not globally and conn_or_curs or None)
 TypeError: argument 2 must be a connection, cursor or None
```
The fix is simply to pass unwrapped connection or cursor.

### CODE REVIEW

File: skywalking/plugins/sw_psycopg2.py [hunk 1]
```diff
@@ -128,3 +128,53 @@ def connect(*args, **kwargs):
...
+
+            return _register_type(c, conn_or_curs)
+
+        try:
+            import psycopg2._ipaddress
+
+            if psycopg2._ipaddress.register_type is _register_type:
+                psycopg2._ipaddress.register_type = register_type
+
+        except Exception:
```
[the remark this answers is not in the document]
Comment (tom-pytel) [reply skywalking/plugins/sw_psycopg2.py:149]: ignore

**Comment by Superskyyy** [comment]
This addresses this https://github.com/apache/skywalking/discussions/8556 problem, right? Thanks for looking into this.

**Comment by tom-pytel** [comment]
> This addresses this [apache/skywalking#8556](https://github.com/apache/skywalking/discussions/8556) problem, right? Thanks for looking into this.

Should, yes.

**Comment by Superskyyy** [comment]
> > This addresses this [apache/skywalking#8556](https://github.com/apache/skywalking/discussions/8556) problem, right? Thanks for looking into this.
> 
> Should, yes.

If you have verified it works for Django version above 3.10 then LGTM. :)

**Comment by tom-pytel** [comment]
> > > This addresses this [apache/skywalking#8556](https://github.com/apache/skywalking/discussions/8556) problem, right? Thanks for looking into this.
> > 
> > 
> > Should, yes.
> 
> If you have verified it works for Django version above 3.10 then LGTM. :)

There is a Django 3.10? I did do a quick test of our fork of this with Django 4.0.5 and it works, waiting on more thorough tests from the test monkeys though.

**Comment by Superskyyy** [comment]
> > > > This addresses this [apache/skywalking#8556](https://github.com/apache/skywalking/discussions/8556) problem, right? Thanks for looking into this.
> > > 
> > > 
> > > Should, yes.
> > 
> > 
> > If you have verified it works for Django version above 3.10 then LGTM. :)
> 
> There is a Django 3.10? I did do a quick test of our fork of this with Django 4.0.5 and it works, waiting on more thorough tests from the test monkeys though.

LOL I forgot Django is already at 4.x. Btw we should bump up the checked versions in plugin supported matrix.
```
support_matrix = {
    'django': {
        '>=3.6': ['3.2'],
        # ">=3.8": ["4.0a1"]  # expected Dec 2021
    }
}
``` change to something like

```
support_matrix = {
    'django': {
        '>=3.6': ['3.2'],
        ">=3.8': ['4.0'] # 4.2 will support py3.11 
    }
}
```

BTW, do you think we should remove the support for 3.6? It has been sunsetted since 6 months ago. We will introduce some advanced features requiring 3.7 up, like the automatic post-fork hook.

**Comment by tom-pytel** [comment]
> LOL I forgot Django is already at 4.x. Btw we should bump up the checked versions in Django plugin supported matrix.

That is for a different PR, this one is about psycopg2.

> BTW, do you think we should remove the support for 3.6? It has been sunsetted since 6 months ago. We will introduce some advanced features requiring 3.7 up, like the automatic post-fork hook.

If you are asking my opinion then don't remove support for something until and unless you need to. There's nothing worse than having your project stop working for no good reason other than an arbitrary date.

**Review by Superskyyy — APPROVED** [review]

**Comment by Superskyyy** [comment]
> That is for a different PR, this one is about psycopg2.
> 

> If you are asking my opinion then don't remove support for something until and unless you need to. There's nothing worse than having your project stop working for no good reason other than an arbitrary date.

Yes, thanks for the explanation. LGTM

**Pull request merged**